src/plugins/extra/typographer.rs

```diff
@@ -0,0 +1,121 @@
+        // These look a little different from the JS implementation because the
+        // regex crate doesn't support look-behind and look-ahead patterns
+        (
+            Regex::new(r"(?m)(?P<pre>^|[^-])(?P<dash>---)(?P<post>[^-]|$)").unwrap(),
+            "$pre\u{2014}$post",
+        ),
+        (
+            Regex::new(r"(?m)(?P<pre>^|\s)(?P<dash>--)(?P<post>\s|$)").unwrap(),
+            "$pre\u{2013}$post",
+        ),
+        (
+            Regex::new(r"(?m)(?P<pre>^|[^-\s])(?P<dash>--)(?P<post>[^-\s]|$)").unwrap(),
+            "$pre\u{2013}$post",
+        ),
+    ])
+});
+static SCOPED_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?i)\((c|tm|r)\)").unwrap());
+static RARE_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\+-|\.\.|\?\?\?\?|!!!!|,,|--").unwrap());
+
+fn replace_abbreviation(input: &str) -> &'static str {
+    match input.to_lowercase().as_str() {
+        "(c)" => "©",
+        "(r)" => "®",
+        "(tm)" => "™",
+        _ => panic!("Got invalid abbreviation '{}'", input),
+    }
+}
+
+pub fn add(md: &mut MarkdownIt) {
+    md.add_rule::<TypographerRule>();
+}
+
+pub struct TypographerRule;
+
+impl Replacer for TypographerRule {
+    fn replace_append(&mut self, caps: &regex::Captures<'_>, dst: &mut String) {
+        dst.push_str(replace_abbreviation(caps.get(0).unwrap().as_str()));
+    }
+}
+
+impl CoreRule for TypographerRule {
+    fn run(root: &mut Node, _: &MarkdownIt) {
+        root.walk_mut(|node, _| {
+            if let Some(mut text_node) = node.cast_mut::<Text>() {
+                if SCOPED_RE.is_match(&text_node.content) {
+                    text_node.content = SCOPED_RE.replace_all(&text_node.content, Self).to_string();
+                }
+                if RARE_RE.is_match(&text_node.content) {
+                    let mut result = text_node.content.to_owned();
+                    for (pattern, replacement) in REPLACEMENTS.iter() {
+                        // This is a bit unfortunate, but since we can't use
+                        // look-ahead and look-behind patterns in the dash
+                        // replacements, the preceding and following characters (pre
+                        // and post in the patterns) become part of the match.
+                        // So a string like "bla-- --foo" would create two
+                        // *overlapping* matches, "a-- " and " --f". But replace_all
+                        // only replaces non-overlapping matches. So we can't do
+                        // this in one single replacement.
+                        // My only consolation here is that this won't happen very
+                        // often in practice, and in any case it's probably good to
+                        // ask whether the patterns match before attempting any
+                        // replacement, since that's supposed to be the cheaper
+                        // operation.
+                        while pattern.is_match(&result) {
```

Comment:
If this becomes a performance problem, we can always replace these regexps with a custom thing (shouldn't be too hard to manually traverse a string and make replacements).

> and in any case it's probably good to ask whether the patterns match before attempting any replacement, since that's supposed to be the cheaper operation.

Is it cheaper to check string twice? I'm not sure. I think you're making it expensive by using `.to_string()` everywhere.

This code for example does the same thing, but doesn't do any extra allocations:

```rs
if RARE_RE.is_match(&text_node.content) {
    let mut result = Cow::Borrowed(text_node.content.as_str());

    for (pattern, replacement) in REPLACEMENTS.iter() {
        if let Cow::Owned(s) = pattern.replace_all(&result, *replacement) {
            result = Cow::Owned(s);

            // do the replacement once more for that overlapping edge case
            if let Cow::Owned(s) = pattern.replace_all(&result, *replacement) {
                result = Cow::Owned(s);
            }
        }
    }

    if let Cow::Owned(s) = result {
        text_node.content = s;
    }
}
```

Disclaimer: I didn't do benchmarks.

Reply:
I also didn't do benchmarks, I'm just going by this section from the [regex crate](https://docs.rs/regex/latest/regex/#pay-for-what-you-use) docs:

> However, it can be significantly more expensive to compute the location of capturing group matches, so it’s best not to do it if you don’t need to.
> 
> Therefore, only use what you need. For example, don’t use find if you only need to test if an expression matches a string. (Use is_match instead.)

When I write that it makes sense to ask first in any case, I am making the (reasonable IMHO but debatable) assumption that _most_ strings we encounter won't contain any replacement candidates, and that for all these inputs the cheaper call to `is_match` saves us the time to make the more expensive call to `replace`. This will of course not hold for all inputs, specifically those that are primarily composed of dashes. :P

Reply:
I do take your point about `to_string` to heart though, that's great feedback, thanks! :)

Reply:
resolved in #4